reasoner/planner/RecursivePorPlanner.java

```diff
@@ -0,0 +1,361 @@
+        // Walk the cycle, backtrack over choices of orders.
+        if (pendingKeys.isEmpty()) {
+            return new Pair<>(new ArrayList<>(chosenSummaries.values()), cyclePlanCost(root, chosenSummaries));
+        }
+
+        CallKey key = pendingKeys.stream().findAny().get();
+        pendingKeys.remove(key);
+        assert !chosenSummaries.containsKey(key); // Should not have been added to pending
+
+        ConjunctionSummary conjunctionSummary = conjunctionSummarizer.conjunctionSummary(key.conjunction);
+        Pair<List<OrderingSummary>, Double> bestPlan = null; // for this branch
+        for (OrderingSummary orderingSummary : candidateOrderings.get(key)) {
+            chosenSummaries.put(key, orderingSummary);
+
+            Set<CallKey> nextPendingKeys = new HashSet<>(pendingKeys);
+            Set<CallKey> triggeredCalls = orderingSummary.cyclicCallsTriggered(conjunctionSummary);
+            iterate(triggeredCalls).filter(call -> !chosenSummaries.containsKey(call)).forEachRemaining(nextPendingKeys::add);
+            Pair<List<OrderingSummary>, Double> newPlan = cyclePlanSearch(root, nextPendingKeys, chosenSummaries);
+            bestPlan = (bestPlan == null || newPlan.second() < bestPlan.second()) ? newPlan : bestPlan;
+
+            chosenSummaries.remove(key);
+        }
+        assert bestPlan != null;
+        return bestPlan;
+    }
+
+    private void initializeOrderingDependencies(ConjunctionSummary conjunctionSummary, List<Resolvable<?>> ordering, Set<Variable> inputBounds) {
+        Set<Variable> currentBounds = new HashSet<>(inputBounds);
+        for (Resolvable<?> resolvable : ordering) {
+            Set<Variable> resolvableBounds = iterate(consideredVariables(resolvable)).filter(currentBounds::contains).toSet();
+            initializeDependencies(conjunctionSummary, resolvable, resolvableBounds);
+            currentBounds.addAll(consideredVariables(resolvable));
+        }
+    }
+
+    private void initializeDependencies(ConjunctionSummary conjunctionSummary, Resolvable<?> resolvable, Set<Variable> resolvableBounds) {
+        if (resolvable.isConcludable()) {
+            Set<ResolvableConjunction> cyclicDependencies = conjunctionSummary.cyclicDependencies(resolvable.asConcludable());
+            for (Rule rule : logicMgr.applicableRules(resolvable.asConcludable()).keySet()) {
+                ResolvableConjunction ruleConjunction = rule.condition().conjunction();
+                recursivelySummarizeOrderings(new CallKey(ruleConjunction, resolvableBounds));
+                if (!cyclicDependencies.contains(ruleConjunction)) {
+                    plan(ruleConjunction, resolvableBounds); // Acyclic dependencies can be fully planned
+                }
+            }
+        }
+    }
+
```

Comment:
The initialise methods generate candidate orderings for each triggered (conjunction, bounds) and prune the redundant, suboptimal ones.